File: solutions/c/binary-search/2/binary_search.c
#include "binary_search.h"

const int* binary_search(int target, const int* sorted_items, size_t length) {
    if (length == 0) return NULL;

    size_t low = 0;
    size_t high = length - 1;

    while (low <= high) {
        size_t mid = low + (high - low) / 2;

        if (sorted_items[mid] == target)
            return &sorted_items[mid];

        if (target < sorted_items[mid]) {
            if (mid == 0) return NULL;
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }
    return NULL;
}

File: solutions/elixir/binary-search/1/lib/binary_search.ex
defmodule BinarySearch do

  def search({}, _target), do: :not_found

  def search(numbers, target) when is_tuple(numbers) do
    index_of(numbers, target, 0, tuple_size(numbers) - 1)
  end

  defp index_of(_numbers, _target, lo, hi) when lo > hi, do: :not_found

  defp index_of(numbers, target, lo, hi) do
    mid = div(lo + hi, 2)

    case elem(numbers, mid) do
      ^target -> {:ok, mid}
      too_small when too_small < target -> index_of(numbers, target, mid + 1, hi)
      _too_large -> index_of(numbers, target, lo, mid - 1)
    end
  end
  
end

